src/Raven.Client/Documents/BulkInsert/BulkInsertOperation.cs

```diff
@@ -263,6 +272,60 @@ public BulkInsertOperation(string database, IDocumentStore store, BulkInsertOpti
 
             _generateEntityIdOnTheClient = new GenerateEntityIdOnTheClient(_requestExecutor.Conventions,
                 entity => AsyncHelpers.RunSync(() => _requestExecutor.Conventions.GenerateDocumentIdAsync(database, entity)));
+
+            _streamLock = new AsyncReaderWriterLock();
+            _lastWriteToStream = DateTime.UtcNow;
+            _timer = new Timer(HandleHeartbeat,
+                new WeakReference<BulkInsertOperation>(this),
+                HeartbeatCheckInterval,
+                HeartbeatCheckInterval);
+        }
+
+        private static void HandleHeartbeat(object state)
+        {
+            var weakReference = (WeakReference<BulkInsertOperation>)state;
+            if (weakReference.TryGetTarget(out BulkInsertOperation bulkInsert) == false)
+                return;
+            AsyncHelpers.RunSync( () => SendHeartBeatAsync(bulkInsert));
+        }
+
+        private static async Task SendHeartBeatAsync(BulkInsertOperation bulkInsert)
+        {
+            if (DateTime.UtcNow.Ticks - bulkInsert._lastWriteToStream.Ticks < HeartbeatCheckInterval.Ticks)
+                return;
+
+            using (await bulkInsert._streamLock.WriterLockAsync().ConfigureAwait(false))
```

Comment:
why bother trying to lock and wait?
Instead, there is a `TryLock`, no?

`AsyncReaderWriterLock` doesn't support that, but what is the reasoning behind not using `SempahoreSlim` ?